neo/Ledger/BlockChain.State.cs

```diff
@@ -0,0 +1,162 @@
+            if (!stateRoot.Verify(currentSnapshot)) return RelayResultReason.Invalid;
+            if (stateRootCache.ContainsKey(stateRoot.Index)) return RelayResultReason.AlreadyExists;
+            if (stateRoot.Index > Height || (stateRoot.Index > StateHeight + 1 && stateRoot.Index != StateRootEnableIndex))
+                return RelayResultReason.Succeed;
+
+            using (Snapshot snapshot = GetSnapshot())
+            {
+                while (stateRoot.Index <= Height)
+                {
+                    if (PersistCnStateRoot(stateRoot) == StateRootVerifyFlag.Invalid)
+                        break;
+
+                    if (stateRoot.Index + 3 > HeaderHeight)
+                    {// TODO remove +3 and use LocalNode.RelayDirectly
+                        system.LocalNode.Tell(new LocalNode.SendDirectly { Inventory = stateRoot });
+                    }
+
+                    stateRootCache.Remove(stateRoot.Index);
+                    if (!stateRootCache.TryGetValue(stateRoot.Index + 1, out stateRoot)) break;
+                }
+            }
+            return RelayResultReason.Succeed;
+        }
+
+        private void OnStateRoots(StateRoot[] sts)
+        {
+            foreach (var state_root in sts)
+            {
+                var result = OnNewStateRoot(state_root);
+                if (result != RelayResultReason.Succeed)
+                    break;
+            }
+            system.TaskManager.Tell(new TaskManager.StateRootTaskCompleted(), Sender);
+        }
+
+        private void OnImportRoots(IEnumerable<StateRoot> roots)
+        {
+            foreach (StateRoot stateRoot in roots)
+            {
+                if (stateRoot.Index < Math.Max(StateHeight, StateRootEnableIndex)) continue;
+                if (stateRoot.Index != Math.Max(StateHeight + 1, StateRootEnableIndex))
+                    throw new InvalidOperationException();
+
+                if (PersistCnStateRoot(stateRoot) == StateRootVerifyFlag.Invalid)
+                    break;
+            }
+            Sender.Tell(new ImportCompleted());
+        }
+
+        private StateRootVerifyFlag PersistCnStateRoot(StateRoot stateRoot)
+        {
+            using (Snapshot snapshot = GetSnapshot())
+            {
+                var localState = snapshot.StateRoots.GetAndChange(stateRoot.Index);
+                if (localState.StateRoot.Root == stateRoot.Root && localState.StateRoot.PreHash == stateRoot.PreHash)
+                {
+                    HashIndexState hashIndexState = snapshot.StateRootHashIndex.GetAndChange();
+                    hashIndexState.Index = stateRoot.Index;
+                    hashIndexState.Hash = stateRoot.Hash;
+                    localState.StateRoot = stateRoot;
+                    localState.Flag = StateRootVerifyFlag.Verified;
+                }
+                else
+                {
+                    localState.Flag = StateRootVerifyFlag.Invalid;
```

Comment:
Any log, how we will know that this happens? stop the chain?

Reply:
Chain still go ahead. Stop the StateHeight.

Reply:
What is the sense of the state validation if we don't know if it's valid? how will be the user awarned if the node have different states?

Reply:
neo-cli should display the state height. If state height is not equal block height, user should upgrade neo node, or something wrong.

Reply:
I think that this should be optional, I think that the blockchain should not sync if it's wrong.

Reply:
Maybe this should be decided by clients, like neo-cli and neo-gui.